File: ts/D9.ts
import { readFileSync } from 'fs';
import { timeAndPrint } from './time-and-print';

function parseInput(): number[] {
    const numbers = [];
    const content = readFileSync('../inputs/D9.input','utf8');
    for (let line of content.split("\r\n")) {
        numbers.push(+line);
    }
    return numbers;
}

function findWeakness(numbers: number[]): number {
    for (let i = 25; i < numbers.length; i++) {
        let found = false;
        for (let j = i - 25; j < i; j++) {
            for (let k = j + 1; k < i; k++) {
                if (numbers[j] + numbers[k] == numbers[i]) {
                    found = true;
                    break;
                }
            }
            if (found) break;
        }

        if (!found) return numbers[i];
    }
}

function p1() {
    return findWeakness(parseInput());
}

function p2() {
    const numbers = parseInput();
    const weakness = findWeakness(numbers);

    let L = 0, R = 1, sum = numbers[0] + numbers[1];

    while (sum != weakness) {
        if (sum > weakness) {
            sum -= numbers[L];
            L++;
        } else {
            R++;
            sum += numbers[R];
        }
    }

    const range = numbers.slice(L, R + 1);
    const min = Math.min(...range);
    const max = Math.max(...range);
    return min + max;
}

timeAndPrint(p1);
timeAndPrint(p2);

File: py/D9.py

def parseInput():
    file = open("../inputs/D9.input", "r")
    return list(map(int, file.read().splitlines()))

def findWeakness(numbers):
    for i in range(25, len(numbers)):
        found = False
        for j in range(i - 25, i):
            for k in range(j + 1, i):
                if (numbers[j] + numbers[k] == numbers[i]):
                    found = True
                    break
            if found: break
        if not found: return numbers[i]

def p1():
    return findWeakness(parseInput())

def p2():
    numbers = parseInput()
    weakness = findWeakness(numbers)
    L = 0
    R = 1
    sum = numbers[0] + numbers[1]
    while sum != weakness:
        if sum > weakness:
            sum -= numbers[L]
            L += 1
        else:
            R += 1
            sum += numbers[R]

    return max(numbers[L:R+1]) + min(numbers[L:R+1])

print(p1())
print(p2())
